Firestore/core/test/unit/model/values_test.cc

```diff
@@ -0,0 +1,331 @@
+    FieldValue fv = testutil::WrapObject((key_value_pairs)...);
+    return serializer.EncodeFieldValue(fv);
+  }
+
+  template <typename... Args>
+  google_firestore_v1_Value WrapArray(Args... values) {
+    std::vector<model::FieldValue> contents{Value(values)...};
+    FieldValue fv = FieldValue::FromArray(std::move(contents));
+    return serializer.EncodeFieldValue(fv);
+  }
+
+  google_firestore_v1_Value WrapReference(DatabaseId database_id,
+                                          DocumentKey key) {
+    google_firestore_v1_Value result{};
+    result.which_value_type = google_firestore_v1_Value_reference_value_tag;
+    result.reference_value =
+        serializer.EncodeResourceName(database_id, key.path());
+    return result;
+  }
+
+  google_firestore_v1_Value WrapServerTimestamp(
+      const model::FieldValue& input) {
+    // TODO(mrschmidt): Replace with EncodeFieldValue encoding when available
+    return WrapObject("__type__", "server_timestamp", "__local_write_time__",
+                      input.server_timestamp_value().local_write_time());
+  }
+
+  template <typename... Args>
+  void Add(std::vector<std::vector<google_firestore_v1_Value>>& groups,
+           Args... values) {
+    std::vector<google_firestore_v1_Value> group{(values)...};
+    groups.emplace_back(group);
+  }
+
+  void VerifyEquality(std::vector<google_firestore_v1_Value>& left,
+                      std::vector<google_firestore_v1_Value>& right,
+                      bool expected_equals) {
+    for (const auto& val1 : left) {
+      for (const auto& val2 : right) {
+        EXPECT_EQ(expected_equals, Values::Equals(val1, val2))
+            << "Equality check failed for '" << Values::CanonicalId(val1)
+            << "' and '" << Values::CanonicalId(val2) << "' (expected "
+            << expected_equals << ")";
+      }
+    }
+  }
+
+  void VerifyOrdering(std::vector<google_firestore_v1_Value>& left,
+                      std::vector<google_firestore_v1_Value>& right,
+                      ComparisonResult expected_result) {
+    for (const auto& val1 : left) {
+      for (const auto& val2 : right) {
+        EXPECT_EQ(expected_result, Values::Compare(val1, val2))
+            << "Order check failed for '" << Values::CanonicalId(val1)
+            << "' and '" << Values::CanonicalId(val2) << "' (expected "
+            << static_cast<int>(expected_result) << ")";
+        EXPECT_EQ(util::ReverseOrder(expected_result),
+                  Values::Compare(val2, val1))
+            << "Reverse order check failed for '" << Values::CanonicalId(val1)
+            << "' and '" << Values::CanonicalId(val2) << "' (expected "
+            << static_cast<int>(util::ReverseOrder(expected_result)) << ")";
+      }
+    }
+  }
+
+  void VerifyCanonicalId(const google_firestore_v1_Value& value,
+                         const std::string& expected_canonical_id) {
+    const std::string& actual_canonical_id = Values::CanonicalId(value);
```

Comment:
Does this const reference bind to a temporary? If yes, please replace it with a value.